File: src/main/java/play/pluv/feed/application/FeedReader.java
package play.pluv.feed.application;

import static java.util.Comparator.comparing;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;
import play.pluv.feed.domain.Feed;
import play.pluv.feed.domain.FeedBookmark;
import play.pluv.feed.domain.repository.FeedBookmarkRepository;
import play.pluv.feed.domain.repository.FeedRepository;
import play.pluv.member.domain.Member;

@Component
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class FeedReader {

  private final FeedRepository feedRepository;
  private final FeedBookmarkRepository feedBookmarkRepository;

  public List<Feed> findAll() {
    return feedRepository.findAll().stream()
        .sorted(comparing(Feed::getCreatedAt).reversed())
        .sorted(comparing(feed -> feed.getId() <= 116))
        .toList();
  }

  public List<Feed> findBookmarkedFeeds(final Member member) {
    return feedBookmarkRepository.findByMemberIdWithJoin(member.getId()).stream()
        .map(FeedBookmark::getFeed)
        .toList();
  }

  public Feed findFeed(final Long feedId) {
    return feedRepository.readById(feedId);
  }

  public Boolean isBookMarked(final Feed feed, final Long memberId) {
    return feedBookmarkRepository.existsByMemberIdAndFeed(memberId, feed);
  }
}

File: src/main/kotlin/play/pluv/feed/application/FeedReader.kt
package play.pluv.feed.application

import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import play.pluv.feed.domain.Feed
import play.pluv.feed.domain.repository.FeedBookmarkRepository
import play.pluv.feed.domain.repository.FeedRepository
import play.pluv.member.domain.Member

@Component
@Transactional(readOnly = true)
class FeedReader(
    private val feedRepository: FeedRepository,
    private val feedBookmarkRepository: FeedBookmarkRepository
) {

    fun findAll(): List<Feed> {
        return feedRepository.findAll()
            .sortedByDescending { it.createdAt }
    }

    fun findBookmarkedFeeds(member: Member): List<Feed> {
        return feedBookmarkRepository.findByMemberIdWithJoin(member.identifier)
            .sortedByDescending { it.createdAt }
            .map { it.feed }
    }

    fun findFeed(feedId: Long): Feed {
        return feedRepository.readById(feedId)
    }

    fun isBookMarked(feed: Feed, memberId: Long): Boolean {
        return feedBookmarkRepository.existsByMemberIdAndFeed(memberId, feed)
    }
}
